Refactor and optimize, implement Neighbours

colega opened the pull request.

colega commented:
Sorry, I wanted to merge this to our fork

dim commented on v3/tile.go:
```diff
@@ -0,0 +1,190 @@
+	return Tile{X: x, Y: y, Level: level}, nil
+}
+
+func (t Tile) Neighbours() []Tile {
```
apart from this, is there anything else fundamentally new in your PR? doesn't seem to be difficult to merge

colega commented:
Let me try it...
The big change is that I changed lots of logic to integers, removed unnecessary projections, and removed exposing the underneath projection to end-users
